[Comb] Canonicalize mux with constant inputs

richardxia opened the pull request:
Fixes the second half of #1624.

This simplifies mux expressions that have constant values for both the true and false cases by pushing the mux condition down to the individual bits of the inputs, combining the results with a concat op. This takes advantage of the simpler folds/canonicalization patterns that have already been implemented for muxes.

Separately, I also added a canonicalization pattern for the special case of `mux(cond, 0, 1'h1)`, which is equivalent to `~cond`. This helps with matching other patterns that look for bitwise not operations.

In the simplest case, it should simplify expressions like `mux(cond, 2'h2, 2'h0) -> concat(mux(cond, 1'h1, 1'h0), 1'h0) -> concat(cond, 1'h0)`. More generally, the end result should be the concatenation of one of four possible values: 0, 1, `cond`, or `~cond`.

I wasn't sure if applying this transformation on all mux ops that have constant inputs was being too aggressive or too conservative, since I don't have a great sense of what is "simpler," so I was hoping to get some guidance here. I can see that this transformation might make some mux expressions even more complicated (for example `mux(cond, 3'h6, 0) -> concat(cond, cond, 0)`), so I could restrict this something even stricter than "any mux op with constant inputs". On the flip side, I can also see that restricting it to constants may cause us to miss out on cases where we still be able to simplify the expression.

One somewhat negative effect of this transformation can be seen in the Comb canonicalization tests, specifically in the SevenSegmentDecoder test case. Here it looks like the result of my changes causes several duplicates of `comb.xor %0, %true : i1` in the output. These go away when `-cse` is run, but I don't know if that's something we want to turn on for these tests.

lattner commented:
This is really awesome, thank you for tackling this!

> I wasn't sure if applying this transformation on all mux ops that have constant inputs was being too aggressive or too conservative, since I don't have a great sense of what is "simpler," so I was hoping to get some guidance here. 

I agree, I think this is the crux of the question.  [This is the section](https://github.com/llvm/circt/blob/main/docs/RationaleHW-SV.md#cost-model) of the rationale that we've started to discuss this, but as you say, we need to talk about this specific case more.

In my opinion, at the "HW" level of abstraction, we shouldn't drop to literally bit level modeling egregiously, even though it may open other simplifications.    Such a thing will bloat the IR with lots of operations, and will slow down verilator and other things quite a bit.

The most conservative thing to do is to implement the general pass and then gate it with a bunch of heuristics in which it is "obviously good", such as when the simplification drops the mux condition into exactly one bit and the rest turns constant.  This is moving complexity around, but not exploding the size of the IR.

I'll take a detailed look at your patch now.

lattner commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -1684,6 +1684,18 @@ static bool foldMuxChain(MuxOp rootMux, bool isFalseSide,
+        value2.isAllOnes()) {
+      auto one = rewriter.create<hw::ConstantOp>(op.getLoc(), APInt(1, 1));
+      rewriter.replaceOpWithNewOp<XorOp>(op, op.cond(), one);
+      return success();
```
This logic LGTM, but please merge it into the stuff below so the "trueValue" is only matched to a constant once.  One additional tweak: can you reuse `op.trueValue()` (which you known is a single bit one value) rather than creating a new "one" constantop?

richardxia replied:
Sounds good, and good call on reusing `op.trueValue()`.

lattner commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -1707,6 +1719,28 @@ LogicalResult MuxOp::canonicalize(MuxOp op, PatternRewriter &rewriter) {
+  // the inputs and attempt to fold each bitwise mux.
+  // E.g. mux(a, 2, 0) -> concat(mux(a, 1, 0), 0) -> concat(a, 0)
+  if (matchPattern(op.trueValue(), m_RConstant(value)) &&
+      value.getBitWidth() > 1) {
```
Similar comment: please conserve matchPattern calls by merging these things together a bit.

lattner commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -1707,6 +1719,28 @@ LogicalResult MuxOp::canonicalize(MuxOp op, PatternRewriter &rewriter) {
+        operands.push_back(rewriter.createOrFold<comb::MuxOp>(
+            op.getLoc(), op.cond(), trueValue, falseValue));
+      }
+      rewriter.replaceOpWithNewOp<ConcatOp>(op, op.getType(), operands);
```
Per your comment, I think we should start by only doing this when we aren't carving it up into too many pieces.  How about we start with "exactly one bit different between the two constants"?  You can check for this with something like `(trueValue^falseValue).isPowerOfTwo()`.

Given that, your concat will always have two or three chunks - plz build it that way, instead of building it bitwise and relying on the concat to merge it back.  This would save a lot of compile time, particularly with i128's or whatever.

richardxia replied:
> How about we start with "exactly one bit different between the two constants"?

Yep, this sounds like a good starting point.

> Given that, your concat will always have two or three chunks - plz build it that way, instead of building it bitwise and relying on the concat to merge it back.

Definitely, will do.

lattner reviewed (commented):
Thank you for tackling this Richard!

richardxia commented:
I believe I've addressed all the comments, including reimplementing the main logic here to operate directly on the 2-3 known chunks instead of iterating bitwise on the operands. Thanks again for the feedback!

lattner commented on lib/Dialect/Comb/CombFolds.cpp:
```diff
@@ -1684,11 +1684,64 @@ static bool foldMuxChain(MuxOp rootMux, bool isFalseSide,
+          // the 1 in the representation of ~ using xor.
+          rewriter.replaceOpWithNewOp<XorOp>(op, op.cond(), op.falseValue());
+          return success();
+        }
```
This isn't simplifying the single bit `mux(cond, 0, x)` case.  Can that / should that turn into an `and(~cond, x)`?

lattner replied:
One nice thing about handling all the 1-bit cases is that the "Else" on line 1703 would go away since this "if" would be total.  This would make it easier to understand the logic here by reducing nesting.

richardxia replied:
Nice, I like that this makes the `else` go away. I've made this change in my latest commit. The only drawback here is that I think I can no longer rely on the trick of using the falseValue being a known one-bit `1`, meaning that I have to create a ConstantOp for the 1 in the Xor again, but that seems worth it in order to make this much more general.

I also noticed that this applies to the other conditional branch lower in the function, where we only pattern match on the falseValue being constant, so I've made the equivalent change there. This should mean that we're now able to reduce all muxes with single-bit inputs to some simpler expression with `and`, `or`, or `not`s. I've added several more FileCheck tests to double check that each of these cases is being handled properly since I don't completely trust myself to not make a bit flip error here.

lattner reviewed (approved):
This is really great, nice work!

lattner reviewed (approved):
This is really great, nice and simple, I love it.  Thanks!
